docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+* **Completed with an exception:** If `sum_values` raised an exception while processing, then re-raise that same exception, causing it to be displayed to the user in the `sum` output.
+
+It's not necessary to memorize these states.
+Just remember that `sum_values.result()` is a reactive, synchronous method that knows how to do right thing based on the state of the extended task.
+
+### Other features of extended tasks
+
+#### Cancel a running task
+
+Although it's not shown in the example above, you can also cancel an extended task by calling the `cancel()` method (for example, `sum_values.cancel()`).
+This will attempt to cancel the asyncio task that is running the extended task, and will cause `sum_values.result()` to raise a silent exception.
+
+Calling `sum_values.cancel()` on a task that isn't running will have no effect.
+
+#### Multiple invocations
+
+An extended task can run concurrently to reactive code and to other extended tasks--that's its whole purpose.
+However, a single extended task object cannot run itself multiple times concurrently. If you call `sum_values()` while it is already running, it will enqueue the new invocation and run it after the first one completes.
+
+This is often not the behavior you want, especially if the task takes a long time to complete.
+A user may accidentally click an action button twice, or they may click it again because they think the first click didn't work.
+To prevent this, use replace `ui.input_action_button` with `ui.input_task_button`, which will automatically prevent subsequent clicks until the task completes.
+
+### Executing on a different thread/process
+
+Extended task objects run their tasks using [`asyncio.create_task`](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task), which means that they will run on the same thread as the rest of the application.
+This is fine for logic that is truly asynchronous, like making a network request via an asyncio-aware library like [aiohttp](https://docs.aiohttp.org/en/stable/), but it's not ideal for CPU-bound tasks or when performing I/O synchronously.
+Because CPU-bound or synchronous tasks will block the main thread, we're back to where we started: the rest of the application cannot proceed until the task completes.
+
+Fortunately, we can rely on Python's built-in support for running asyncio tasks on different threads or processes.
+
+::: {.callout-note}
+The examples below work well in Shiny Core, but a naive port to Shiny Express will not work as well.
+The `ThreadPoolExecutor` and `ProcessPoolExecutor` objects need to be created as module-level variables, not as session-level variables, since we'd ideally like to pool resources across all sessions.
+:::
+
+::: {.callout-note}
+`ProcessPoolExecutor` is not available in Shinylive (e.g. WASM mode).
+`ThreadPoolExecutor` is available in Shinylive and appears to work, but doesn't: it actually performs all of its work on the main thread.
+:::
```

Comment:
The shared module concept helps, but not enough--the pools cannot be cleaned up when the app shuts down, both because 1) there's no `app` object to call `app.on_shutdown` on, and 2) the shared module concept wouldn't give you access to the app object even if there was one.

We can definitely devise a way to do this in the future though.